src/Microsoft.FeatureManagement/ConfigurationFeatureDefinitionProvider.cs

```diff
@@ -109,97 +137,77 @@ public Task<FeatureDefinition> GetFeatureDefinitionAsync(string featureName)
             {
                 _definitions.Clear();
             }
 
-            //
-            // Iterate over all features registered in the system at initial invocation time
-            foreach (IConfigurationSection featureSection in GetFeatureDefinitionSections())
-            {
-                string featureName = GetFeatureName(featureSection);
+            IEnumerable<IConfigurationSection> microsoftFeatureManagementSection = GetFeatureDefinitionSections(MicrosoftFeatureManagementFields.FeatureManagementSectionName);
 
-                if (string.IsNullOrEmpty(featureName))
+            if (microsoftFeatureManagementSection != null)
+            {
+                foreach (IConfigurationSection featureSection in microsoftFeatureManagementSection)
                 {
-                    continue;
-                }
+                    string featureName = featureSection[MicrosoftFeatureManagementFields.Id];
 
-                //
-                // Underlying IConfigurationSection data is dynamic so latest feature definitions are returned
-                FeatureDefinition definition = _definitions.GetOrAdd(featureName, (_) => ReadFeatureDefinition(featureSection));
+                    if (string.IsNullOrEmpty(featureName))
+                    {
+                        continue;
+                    }
 
-                //
-                // Null cache entry possible if someone accesses non-existent flag directly (IsEnabled)
-                if (definition != null)
-                {
-                    yield return definition;
+                    //
+                    // Underlying IConfigurationSection data is dynamic so latest feature definitions are returned
+                    FeatureDefinition definition = _definitions.GetOrAdd(featureName, (_) => ParseMicrosoftFeatureDefinition(featureSection));
+
+                    if (definition != null)
+                    {
+                        yield return definition;
+                    }
                 }
             }
-        }
 
-        private void EnsureInit()
-        {
-            if (_initialized == 0)
-            {
-                IConfiguration MicrosoftFeatureManagementConfigurationSection = _configuration
-                    .GetChildren()
-                    .FirstOrDefault(section =>
-                        string.Equals(
-                            section.Key,
-                            MicrosoftFeatureManagementFields.FeatureManagementSectionName,
-                            StringComparison.OrdinalIgnoreCase));
+            bool rootConfigurationFallback = microsoftFeatureManagementSection == null && RootConfigurationFallbackEnabled;
 
-                bool hasMicrosoftFeatureManagementSchema = MicrosoftFeatureManagementConfigurationSection != null;
+            IEnumerable<IConfigurationSection> dotnetFeatureManagementSection = GetDotnetFeatureManagementSection(rootConfigurationFallback);
```

Comment:
I think it would simplify things a bit if the root configuration fallback was calculated inside of `GetDotnetFeatureManagementSections`